File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpRangeUtil.cs
using JetBrains.Annotations;
using JetBrains.DocumentModel;
using JetBrains.ReSharper.Psi;
using Microsoft.FSharp.Compiler;
using Column = JetBrains.Util.dataStructures.TypedIntrinsics.Int32<JetBrains.DocumentModel.DocColumn>;
using Line = JetBrains.Util.dataStructures.TypedIntrinsics.Int32<JetBrains.DocumentModel.DocLine>;

namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Util
{
  public static class FSharpRangeUtil
  {
    public static int GetDocumentOffset([NotNull] this IDocument document, Line line, Column column)
    {
      return document.GetLineLength(line) >= column
        ? document.GetOffsetByCoords(new DocumentCoords(line, column))
        : document.GetLineEndOffsetNoLineBreak(line);
    }

    public static int GetDocumentOffset([NotNull] this IDocument document, int line, int column)
    {
      return document.GetDocumentOffset((Line) line, (Column) column);
    }

    public static TreeOffset GetTreeOffset([NotNull] IDocument document, Line line, Column column)
    {
      return document.GetLineLength(line) >= column
        ? new TreeOffset(document.GetOffsetByCoords(new DocumentCoords(line, column)))
        : TreeOffset.InvalidOffset;
    }

    public static TreeOffset GetTreeStartOffset([NotNull] this IDocument document, Range.range range)
    {
      return GetTreeOffset(document, range.GetStartLine(), range.GetStartColumn());
    }

    public static TreeOffset GetTreeEndOffset([NotNull] this IDocument document, Range.range range)
    {
      return GetTreeOffset(document, range.GetEndLine(), range.GetEndColumn());
    }

    public static Line GetStartLine(this Range.range range)
    {
      // FCS lines are 1-based
      return (Line) (range.StartLine - 1);
    }

    public static Line GetEndLine(this Range.range range)
    {
      // FCS lines are 1-based
      return (Line) (range.EndLine - 1);
    }

    public static Column GetStartColumn(this Range.range range)
    {
      return (Column) range.StartColumn;
    }

    public static Column GetEndColumn(this Range.range range)
    {
      return (Column) range.EndColumn;
    }

    public static Range.pos GetPos([NotNull] this IDocument document, int offset, int lineShift = 0,
      int columnShift = 0)
    {
      // FCS lines are 1-based
      var coords = document.GetCoordsByOffset(offset);
      return Range.mkPos((int) coords.Line + 1, (int) coords.Column);
    }

    public static Range.pos GetPos(this DocumentCoords coords, int columnShift = 0)
    {
      // FCS lines are 1-based
      return Range.mkPos((int) coords.Line + 1, (int) coords.Column + columnShift);
    }

    public static int GetOffset([NotNull] this IDocument document, Range.pos pos)
    {
      return GetDocumentOffset(document, (Line) (pos.Line - 1), (Column) pos.Column);
    }

    public static bool Contains(this Range.range range, DocumentCoords coords) =>
      range.StartLine - 1 >= (int) coords.Line && range.EndLine - 1 <= (int) coords.Line &&
      range.StartColumn >= (int) coords.Column && range.EndColumn <= (int) coords.Column;
  }
}

File: ReSharper.FSharp/src/FSharp.Common/src/Util/FSharpRangeUtil.fs
[<AutoOpen; Extension>]
module JetBrains.ReSharper.Plugins.FSharp.Util.FSharpRangeUtil

open FSharp.Compiler.Text
open JetBrains.Annotations
open JetBrains.DocumentModel
open JetBrains.ReSharper.Psi
open JetBrains.Util

// FCS lines are 1-based.

[<Extension; CompiledName("ToDocumentCoords")>]
let getDocumentCoords (pos: Position) =
    DocumentCoords(docLine (pos.Line - 1), docColumn pos.Column)

[<Extension; CompiledName("GetDocumentOffset")>]
let getDocumentOffset ([<NotNull>] document: IDocument) (coords: DocumentCoords) =
    if document.GetLineLength(coords.Line) >= coords.Column then
        document.GetOffsetByCoords(coords)
    else
        document.GetLineEndOffsetNoLineBreak(coords.Line)

[<Extension; CompiledName("GetTreeOffset")>]
let getTreeOffsetByCoords ([<NotNull>] document: IDocument) (coords: DocumentCoords) =
    if document.GetLineLength(coords.Line) >= coords.Column then
        TreeOffset(document.GetOffsetByCoords(coords))
    else
        TreeOffset.InvalidOffset

[<Extension; CompiledName("GetTreeOffset")>]
let getTreeOffsetByPos ([<NotNull>] document) pos =
    let coords = getDocumentCoords pos
    getTreeOffsetByCoords document coords

[<Extension; CompiledName("GetTreeStartOffset")>]
let getTreeStartOffset ([<NotNull>] document) (range: range) =
    getTreeOffsetByPos document range.Start

[<Extension; CompiledName("GetTreeEndOffset")>]
let getTreeEndOffset ([<NotNull>] document) (range: range) =
    getTreeOffsetByPos document range.End

[<Extension; CompiledName("GetTreeTextRange")>]
let getTreeTextRange (document: IDocument) (range: range) =
    let startOffset = getTreeStartOffset document range
    let endOffset = getTreeEndOffset document range
    TreeTextRange(startOffset, endOffset)


[<Extension; CompiledName("ToPos")>]
let getPosFromCoords (coords: DocumentCoords) =
    let line = int coords.Line + 1
    let column = int coords.Column
    Position.mkPos line column

[<Extension; CompiledName("GetPos")>]
let getPosFromOffset (document: IDocument) offset =
    let coords = document.GetCoordsByOffset(offset)
    getPosFromCoords coords

[<Extension; CompiledName("GetPos")>]
let getPosFromDocumentOffset (offset: DocumentOffset) =
    getPosFromOffset offset.Document offset.Offset

[<Extension; CompiledName("GetOffset")>]
let getPosOffset ([<NotNull>] document) pos =
    let coords = getDocumentCoords pos
    getDocumentOffset document coords

[<Extension; CompiledName("GetStartOffset")>]
let getStartOffset ([<NotNull>] document) (range: range) =
    getPosOffset document range.Start

[<Extension; CompiledName("GetEndOffset")>]
let getEndOffset ([<NotNull>] document) (range: range) =
    getPosOffset document range.End

[<Extension; CompiledName("ToDocumentRange")>]
let ofFileDocumentRange (documentRange: DocumentRange) (path: FileSystemPath) =
    let startPos = getPosFromDocumentOffset documentRange.StartOffset
    let endPos = getPosFromDocumentOffset documentRange.EndOffset
    Range.mkRange path.FullPath startPos endPos

[<Extension; CompiledName("ToDocumentRange")>]
let ofDocumentRange (documentRange: DocumentRange) =
    ofFileDocumentRange documentRange FileSystemPath.Empty

[<Extension; CompiledName("GetTextRange")>]
let getTextRange (document: IDocument) (range: range) =
    let startOffset = getStartOffset document range
    let endOffset = getEndOffset document range
    TextRange(startOffset, endOffset)

[<Extension; CompiledName("GetDocumentRange")>]
let getDocumentRange (document: IDocument) (range: range) =
    let textRange = getTextRange document range
    DocumentRange(document, textRange)
